emcc.py

```diff
@@ -2630,6 +2630,19 @@ def phase_calculate_system_libraries(state, linker_arguments, linker_inputs, new
   extra_files_to_link += system_libs.calculate(all_linker_inputs, newargs, forced=state.forced_stdlibs)
   linker_arguments.extend(extra_files_to_link)
 
+  if settings.WASMFS and options.embed_files:
+    # wasmfs file embedding is done via emitting C code that contains the data
+    # and code to set them up. we add that as another input, like a system
+    # library, that we compile on the fly here
+    temp_files = shared.configuration.get_temp_files()
+    temp_c = temp_files.get(suffix='.c').name
+    temp_o = unsuffixed(temp_c) + '.o'
+    temp_files.note(temp_o)
+    with open(temp_c, 'w') as f:
+      f.write(package_files(options))
+    shared.check_call([shared.CLANG_CC, temp_c, '-o', temp_o, '-c'] + get_cflags([]))
+    linker_arguments.append(temp_o)
```

Comment:
It looks like `objcopy` is one option, but I'm not sure `llvm-objcopy` is advanced enough yet.

Maybe we can use assembly source code here and call `llvm-mc` to at least avoid calling the C compiler?  I think I'd be OK with that if it try that `llvm-objcopy` is not mature enough yet.

Also, how about moving all of this logic into `file_packager.py` so that all of the logic for generated the object file is encapsulated there?  That way folks who want to run `file_packager.py` themselves can generate and object file to feed to the linker, avoiding this "in-linker" compilation step.

Reply:
Is the issue that running the compiler again at link time is expected to be too slow?

Reply:
In this case its slower than I think it needs to be yes, but the biggest issue for me is that it breaks the normal layering of a toolchain.   I know emcc is historically very magical at link time, so there is precedent, but I'm trying to make it less so where possible.  

I think if the logic is moved out to `file_packager.py` I might feel a little less of a violation since this function could remain unchanged.

Reply:
Moving it to file_packager would be awkward as we'd need to also pass it the relevant cflags. That script is very simple and standalone and not integrated with the main python logic that knows about how emcc compiles or even compilation at all.

But @sbc100 if you want to investigate other options than compiling during the link step that sgtm, I can wait on this PR. I agree this adds a bunch of magic that ideally we'd do without (but I don't see how).

Reply:
What do you think about converting the existing embedding logic to use this method before extending it to wasmfs?  That would avoid ever having two different modes.   I can take a look at doing that if you like?

Reply:
For the old FS I guess we can also embed in a similar way, but I'm not sure how you'd get it to run at the same time as it currently does from the old call from JS. There might be code size and other issues too. Overall I think trying to get that working would be work for little gain? Seems simpler to just not do it and then when we remove the old FS it would be gone.

Reply:
I'd kind of like to do it .. since it would avoid bifurcating the methodology for a potentially unbounded amount of time (since it really not clear now long it could take to deprecate the old FS).  Plus, we would have a great set of tests for it (sine embed file is well tests with the old filesystem).  

Regarding, "There might be code size and other issues too".. wouldn't all those issue apply to wasmfs too?

Reply:
I'm not sure if the issues would affect wasmfs too. Wasmfs is simpler since it's all in wasm. But maybe you have an idea in mind for a good way to do it that avoids issues?

I'm not in a rush to land this, so if you want to look into that feel free to. I'll convert this to a draft.

Reply:
Oh, one specific issue is that with wasmfs we can use internal APIs to embed the file - I have a TODO for that in this PR. That's something we can't do in the old FS. I guess we could EM_ASM out to JS perhaps there..?

Reply:
I got inspired and create a bunch of PRs that cleanup file_packager.py and then one final one that implements preloading via wasm memory: #16050.  I believe this PR will work with both the old FS code and with wasmfs (assuming wasmfs supports FS_createDataFile).  We do followup by removing the dependency on `FS_createDataFile` perhaps but I think its a good way to introduce wasm-embedded files without any other changes.